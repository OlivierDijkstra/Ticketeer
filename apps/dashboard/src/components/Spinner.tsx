import { cva, type VariantProps } from 'class-variance-authority';
import { LoaderCircle } from 'lucide-react';
import * as React from 'react';

import { cn } from '@/lib/utils';

export const variants = {
  variant: {
    default: 'text-accent-foreground',
    primary: 'text-primary-foreground',
  },
  size: {
    default: 'size-8',
    sm: 'size-4',
    lg: 'size-12',
  },
};

const spinnerVariants = cva('grid place-items-center animate-spin', {
  variants,
  defaultVariants: {
    variant: 'default',
    size: 'default',
  },
});

export interface SpinnerProps
  extends React.ButtonHTMLAttributes<SVGSVGElement>,
    VariantProps<typeof spinnerVariants> {
  className?: string;
}

export default function Spinner({
  className,
  variant,
  size,
  ...props
}: SpinnerProps) {
  return (
    <>
      <LoaderCircle
        aria-label='loading'
        aria-busy='true'
        className={cn(spinnerVariants({ className, variant, size }), className)}
        {...props}
      />
      <span className='sr-only'>Loading...</span>
    </>
  );
}
