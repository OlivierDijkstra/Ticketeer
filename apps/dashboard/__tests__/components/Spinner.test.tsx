import { render } from '@testing-library/react';
import { describe, expect, test } from 'vitest';

import Spinner, { variants } from '@/components/Spinner';

type VariantKeys = keyof typeof variants.variant;
const variantArr: VariantKeys[] = Object.keys(
  variants.variant
) as VariantKeys[];

type SizeKeys = keyof typeof variants.size;
const sizeArr: SizeKeys[] = Object.keys(variants.size) as SizeKeys[];

describe('Spinner', () => {
  test('it renders correctly', () => {
    const comp = render(<Spinner />);
    expect(comp.container).toMatchSnapshot();
  });

  variantArr.forEach((variant) => {
    test(`renders correctly with variant: ${variant}`, () => {
      const comp = render(<Spinner variant={variant} />);
      expect(comp.container).toMatchSnapshot();
    });
  });

  sizeArr.forEach((size) => {
    test(`renders correctly with size: ${size}`, () => {
      const comp = render(<Spinner size={size} />);
      expect(comp.container).toMatchSnapshot();
    });
  });
});
