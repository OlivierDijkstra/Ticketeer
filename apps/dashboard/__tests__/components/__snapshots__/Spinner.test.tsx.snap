// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Spinner > it renders correctly 1`] = `
<div>
  <svg
    aria-busy="true"
    aria-label="loading"
    class="lucide lucide-loader-circle grid place-items-center animate-spin text-accent-foreground size-8"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M21 12a9 9 0 1 1-6.219-8.56"
    />
  </svg>
  <span
    class="sr-only"
  >
    Loading...
  </span>
</div>
`;

exports[`Spinner > renders correctly with size: default 1`] = `
<div>
  <svg
    aria-busy="true"
    aria-label="loading"
    class="lucide lucide-loader-circle grid place-items-center animate-spin text-accent-foreground size-8"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M21 12a9 9 0 1 1-6.219-8.56"
    />
  </svg>
  <span
    class="sr-only"
  >
    Loading...
  </span>
</div>
`;

exports[`Spinner > renders correctly with size: lg 1`] = `
<div>
  <svg
    aria-busy="true"
    aria-label="loading"
    class="lucide lucide-loader-circle grid place-items-center animate-spin text-accent-foreground size-12"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M21 12a9 9 0 1 1-6.219-8.56"
    />
  </svg>
  <span
    class="sr-only"
  >
    Loading...
  </span>
</div>
`;

exports[`Spinner > renders correctly with size: sm 1`] = `
<div>
  <svg
    aria-busy="true"
    aria-label="loading"
    class="lucide lucide-loader-circle grid place-items-center animate-spin text-accent-foreground size-4"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M21 12a9 9 0 1 1-6.219-8.56"
    />
  </svg>
  <span
    class="sr-only"
  >
    Loading...
  </span>
</div>
`;

exports[`Spinner > renders correctly with variant: default 1`] = `
<div>
  <svg
    aria-busy="true"
    aria-label="loading"
    class="lucide lucide-loader-circle grid place-items-center animate-spin text-accent-foreground size-8"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M21 12a9 9 0 1 1-6.219-8.56"
    />
  </svg>
  <span
    class="sr-only"
  >
    Loading...
  </span>
</div>
`;

exports[`Spinner > renders correctly with variant: primary 1`] = `
<div>
  <svg
    aria-busy="true"
    aria-label="loading"
    class="lucide lucide-loader-circle grid place-items-center animate-spin text-primary-foreground size-8"
    fill="none"
    height="24"
    stroke="currentColor"
    stroke-linecap="round"
    stroke-linejoin="round"
    stroke-width="2"
    viewBox="0 0 24 24"
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M21 12a9 9 0 1 1-6.219-8.56"
    />
  </svg>
  <span
    class="sr-only"
  >
    Loading...
  </span>
</div>
`;
